### DISCUSSION TITLE
[WIP] RedisRaft 测试相关信息汇总

### DISCUSSION THREAD

**Discussion by happy-v587** [arana-db/kiwi-cpp#61] [General]
FROM：https://github.com/OpenAtomFoundation/pikiwidb/discussions/273

# 目标

- 了解redisraft测试集功能、工作原理等
- 将redisraft测试集适配到 pikiwidb 上
- 针对pikiwidb特性编写专用测试集

参考地址：https://github.com/RedisLabs/redisraft

# 当前进展&未来规划
|功能|进展|其他|
|--|--|--|
|了解redisraft测试集|done|协议 https://github.com/OpenAtomFoundation/pikiwidb/discussions/273#discussioncomment-9109562 ； 使用 https://github.com/OpenAtomFoundation/pikiwidb/discussions/273#discussioncomment-9107680 ； 实现 https://github.com/OpenAtomFoundation/pikiwidb/discussions/273#discussioncomment-9142169|
|适配|doing|见下|

# 适配步骤
- [x] 单机版启动
- [x] 单机版数据写入
- [x] 集群版启动
- [ ] 集群版数据写入
- [ ] 数据正确性校验
- [ ] 故障注入

# 其他

- 大家在调研期间如有有任何疑问欢迎记录并讨论。
- 也鼓励大家分享自己调研结果。便于将信息汇总、也能帮助他人更快速了全方位了解。

**Comment by happy-v587** [comment]
## 关于使用

从官方仓库看测试可以分为两大类：

- 功能测试：即单元测试、集成测试，https://github.com/RedisLabs/redisraft/tree/master/tests
- 正确性测试：即混动工程测试，https://github.com/RedisLabs/redisraft/tree/master/jepsen

## 一、功能测试

基于docker运行测试case

先运行一个容器

```
docker pull debian:buster
docker run -d debian:buster -- tail
docker exec -it 121df5d7a654 bash
```

进入容器后安装一些工具

```
apt-get -y -q update && apt-get install -qqy git wget build-essential virtualenv cmake
```

拉取redis、redisraft代码并编译

```
cd ~
git clone https://github.com/redis/redis.git
cd redis
make
cd ..

git clone https://github.com/RedisLabs/redisraft
cd redisraft 
mkdir build && cd build
cmake .. && make
cd ..
```

运行测试case

```
mkdir -p .env
virtualenv --python=python3 .env
. .env/bin/activate
pip install -r tests/integration/requirements.txt
cd build
make tests
```

测试结果
- 集成测试
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/719a386a-5409-4731-ae4c-443f0783d2d2)
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/374bed15-1ae0-404a-bf31-d3d9951947ee)
- 单元测试
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/6d2a33da-2b12-4858-aea9-dcd5ad6c037c)
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/363cbdc8-fae0-4b92-9dab-05060ed3c232)
- 详细日志
[test.log](https://github.com/OpenAtomFoundation/pikiwidb/files/14996692/test.log)

- 如果要单独跑某个case
```
cd ~/redisraft
pytest tests/integration/test_snapshots.py::test_snapshot_delivery -v --log-cli-level=debug
```

## 二、正确性测试

### 2.1说明

基于Jepsen构建

> Jepsen是Kyle Kingsbury使用Clojure编写的分布式系统测试框架。Clojure是运行在JVM之上的Lisp方言，提供了强大的函数式编程的支持。Leiningen是Clojure的项目构建工具，类似于Maven。事实上，Leiningen底层完全使用Maven的包管理机制，只是Leiningen的构建脚本不是pom.xml，而是project.clj，它本身就是Clojure代码。

### 2.2 在docker中运行

```
cd jepsen/docker
./genkeys.sh
docker-compose build
docker-compose up -d
docker exec -w /jepsen jepsen-control \
    lein run test-all --ssh-private-key /root/.ssh/id_rsa \
        --follower-proxy \
        --time-limit 600 \
        --test-count 50 \
        --concurrency 4n \
        --nemesis partition,pause,kill,member \
        --redis-repo https://github.com/redis/redis \
        --redis-version unstable \
        --raft-repo https://github.com/redislabs/redisraft \
        --raft-version master
```

执行完前四个命令行后会发现启动了如下容器：

<img width="1061" alt="image" src="https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/b10ad6bc-86a7-4db0-bce2-3b4f6e900a05">

查看Dockerfile脚本发现，在 control 容器中你实际是部署了 https://github.com/redislabs/jepsen-redisraft ，这个仓库是使用Clojure编写的、基于jepsen的、用来测试redisraft的工具。

运行起来后可以通过网页 http://127.0.0.1:8080/ 查看

- 运行记录
<img width="1272" alt="image" src="https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/91627422-5fcf-475b-b6cf-4c2d98bf5599">

- 某次运行详细信息（每个文件功能待分析）
<img width="1624" alt="image" src="https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/ec91a382-fb82-4841-99f6-1cf5f0d0e467">

- 得益于jepsen-redisraft，下载、编译、启动、读写数据、验证数据等操作都可以自动化完成
<img width="1791" alt="image" src="https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/f422983b-4439-4d5e-9166-5f80f72de676">
<img src="https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/2686702a-b533-4db8-a232-d0cb405cc4d6">


### 2.3 参考
- https://www.leviathan.vip/2018/05/15/Jepsen%E7%9A%84%E4%BD%BF%E7%94%A8%E4%B8%8E%E7%BA%BF%E6%80%A7%E4%B8%80%E8%87%B4%E6%80%A7/
- https://www.liaoxuefeng.com/article/986637181910336

**Comment by happy-v587** [comment]
## 关于协议

Redisraft LICENSE （SSPLv1.txt、RSALv2.txt）是否会影响我们的选择？

> RSALv2是由Redis Labs创建的协议，用户可以内部使用，但不能将其商业化提供给其他人使用。SSPLv1是由 MongoDB 公司创建的协议，该协议具有传染性，以此构建的软件都必须以相同协议开放源代码。

**Comment by happy-v587** [comment]
## 关于 redisraft 集成测试说明
---

### 两个核心的类

- 通过 `class RedisRaft` 将每个节点的 `init`、`start`、`restart`、`pause`、`join`、`transfer_leader`、`cleanup` 等行为进行管理（通过对 pylib subprocess 、redis cmd 的封装实现）
- 通过 `class Cluster` 将多个 `RedisRaft` 进行管理，包括集群的 `create`、`add_node`、`remove_node`、`pause_leader` 等行为（通过对redis cmd、RedisRaft类的封装实现）

有上面两个给力的 `class` 后，在执行测试case时对集群的生命周期管理就变得非常简单了。

### 测试case的实现

对于每个测试case而言，只需要初始化一个 `cluster`，然后根据需要 `add_node`、`remove_node`；根据需要做 `set key`、`get key` 等操作；根据需要做 `raft COMPACT`等各种操作

- 比如想要测试 raft snapshot 功能是否可以，其中一个case实现是：先启动一个 node 写些数据，然后 raft compact，之后在启动一个新 node，从这个新node查数据，验证是否存在
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/c724b2c9-7a9b-4050-82a9-f07f5b83c8f8)
- 比如想要测试 raft membership 功能是否稳定，其中一个case实现是：先创建一个3节点的集群，删掉1号节点，通过2号节点查询节点数是否变成2
![image](https://github.com/OpenAtomFoundation/pikiwidb/assets/20750625/9da23ddf-d3d0-4a61-962a-c6113c572b2a)

从测试case看，redisraft 主要从上层去关注集群状态是否符合预期。而且这些操作都是通过 raft cmd 去完成

**Comment by happy-v587** [comment]
## 关于 jepsen-redisraft 的测试说明
---
- redis-raft 仓库提供了基于 docker 的运行方式
![image](https://github.com/user-attachments/assets/05e2565c-1199-4bc8-a7a2-7ed8cce139d6)
- 运行的代码存储在 https://github.com/redislabs/jepsen-redisraft
---
### 关于 jepsen-redisraft 仓库的说明

<img width="359" alt="image" src="https://github.com/user-attachments/assets/3facf02b-77b2-4098-b520-46fa5efa8c8b">

0、程序入口

    在 project.clj 中定义了 `:main jepsen.redis.core`
    然后打开 `src/jepsen.readis/core.clj` 文件
    找到文件最后 `defn -main` 即为入口

1、【编译】如何编译 redis、redis-raft 

    在 `defn -main` 中调用了 `defn redis-test`
    在 `defn redis-test` 中调用了 `rdb/redis-raft`
    在 `defn redis-raft` 中实现了 jepsen 要求的
        - `db/DB` 接口：就两个函数 `setup! [db test node]` 和 `teardown! [db test node]`
            - 在 `setup! [db test node]` 中执行了 `install-build-tools!` 、`build-redis!`、`build-redis-raft!`。具体实现类似 bash 逻辑 `mkdir -p xx; git clone xxx; make`

2、【启动集群】如何启动 raft cluster 集群

    还是在 `setup! [db test node]` 中，执行了cmd
        - `redis raft.cluster init` -> `cli! :raft.cluster :init`
        - `redis raft.cluster join` -> `(cli! :raft.cluster :join (str (jepsen/primary test) ":6379")))) `

3、【验证数据】如何生成数据、如何写数据、如何验证数据

    回到 `defn redis-test` 处
    在这里初始化了 workload `let [workload ((workloads (:workload opts)) opts)]`
    而workload的实现在 `src/jepsen.readis/append.clj` 文件的 `defn workload` 函数
    这个函数具体逻辑直接调用的 `jepsen.tests.cycle.append` 测试方法，这个方法包含了数据的生成、校验。
        - 生成逻辑底层其实就是 `elle.list-append.gen`
        - 校验逻辑底层也是 `elle.list-append.check` 的实现

    数据写入是通过定义一个`defrecord Client [conn]`，然后实现`jepsen.client/Client`协议。接口包括：open! close! setup! invoke! teardown!
        - 因为我们要连接的是 redis，所以底层使用的 client 是 https://github.com/taoensso/carmine（类似 go-redis 的存在）
        - 主要逻辑是实现 invoke! ，被读写的数据通过参数传来，所以 invoke! 只负责处理读写操作的执行。在底层就是 `lrange` 和 `rpush`。 最后将读写后得到的数据返回
        - 返回的数据，jepsen会记录到history文件中，最终会通过分析history文件判断数据是否有问题

    关于数据校验可以看 https://github.com/jepsen-io/elle

4、【注入故障】如何注入故障

    jepsen 支持自定义故障，对应接口 `jepsen.nemesis/Nemesis`
        - 比如： 在 `src/jepsen.redis/nemesis.clj` 文件中，`defn member-nemesis` 函数就是 `Nemesis` 接口的一个实现，这个实现的故障是一个节点的 `join` 或者 `leave`
    
    故障的注入也是通过生成器的方式，具体的生成器逻辑可以自己定义。如口在 `src/jepsen.redis/nemesis.clj` 文件的 `defn package` 函数
        - 分为 island-generator 和 mystery-generator
            - island-generator：隔离（网络隔离？还是member leave？通过哪个实现？）
            - mystery-generator：自定义（kill all、kill some、kill part、start、leave、join）

5、【测试框架】put it together

https://blog.csdn.net/FL63Zv9Zou86950w/article/details/117457782

![image](https://github.com/user-attachments/assets/eaa52005-2098-4add-989e-26b18af08e75)
  
以分布式数据库集群为例，Jepsen的工作大致包含以下几步：

      Jepsen在控制节点(Control Node)上作为Clojure程序运行；
      部署需要测试的分布式数据库集群(Distributed System)，确认其工作正常；
      控制节点启动进程，作为分布式数据库节点(DB Node)的SSH客户端；
      生成器(Generator)进程为每个SSH客户端生成具体要执行的读写操作；
      这些SSH客户端将在分布式数据库中执行这些读写操作；
      每个操作从启动到结束的过程都会记录下来；
      当读写操作进行时，克星进程(Nemesis)对分布式数据库进行故障注入，同样也是由生成器来调度和管理；
      测试完成，分布式数据库集群会被销毁。Jepsen使用检查器(Checker)进程对测试过程的历史记录进行分析，并生成最终的图表和报告。

---

### 语法、框架学习

- clojure 语法：https://www.w3cschool.cn/clojure/clojure-j5w81wf2.html
- jepsen 文档：https://jepsen-io.github.io/jepsen/jepsen.cli.html#var-test-all-cmd
- jepsen 训练课：https://jaydenwen123.gitbook.io/zh_jepsen_doc

### 数据验证相关

elle 数据验证工具：https://github.com/jepsen-io/elle
> Elle is a transactional consistency checker for black-box databases.
> As a user, your main entry points into Elle will be elle.list-append/check and elle.rw-register/check.
> Elle has a broad variety of anomalies and consistency models; see elle.consistency-model for their definitions. Not every anomaly is detectable, but we aim for completeness.

elle-cli 支持更多业务模型：https://github.com/ligurio/elle-cli
> is a command-line frontend to transactional consistency checkers for black-box databases. In comparison to Jepsen library it is standalone and language-agnostic tool.

history 文件格式（jepsen使用）: https://github.com/jepsen-io/history
> To analyze these systems, Jepsen uses a history: a totally ordered log of concurrent logical operations.

EDN 文件格式（elle使用）： https://github.com/edn-format/edn
> Extensible Data Notation

JSON -> EDN 文件格式转换：https://github.com/borkdude/jet
> CLI to transform between JSON, EDN, YAML and Transit using Clojure.

### 其他

启动交互式命令行

    # REPL 的环境，即 Read-Eval-Print Loop —- “读取-求值-输出” 循环
    ~/code/cpp/redisraft/jepsen/docker/control/lein repl

调用 list-append/gen 接口

    jepsen.redis.core=> (ns jepsen.redis.core (:require [jepsen.tests.cycle.append :as append]))
    nil
    jepsen.redis.core=> (append/gen)
    (gen/seqone ({:type :invoke, :f :txn, :value [[:append 2 1] [:r 1 nil]]} {:type :invoke, :f :txn, :value [[:append 2 2] [:r 1 nil]]} {:type :invoke, :f :txn, :value [[:r 1 nil] [:append 2 3]]} {:type :invoke, :f :txn, :value [[:append 2 4]]} {:type :invoke, :f :txn, :value [[:append 0 1]]} {:type :invoke, :f :txn, :value [[:r 1 nil]]} {:type :invoke, :f :txn, :value [[:append 2 5]]} {:type :invoke, :f :txn, :value [[:append 2 6] [:r 0 nil]]}))
    
将 gen 结果格式化展示

    {:type :invoke, :f :txn, :value [[:append 2 1] [:r 1 nil]]} 
    {:type :invoke, :f :txn, :value [[:append 2 2] [:r 1 nil]]} 
    {:type :invoke, :f :txn, :value [[:r 1 nil] [:append 2 3]]} 
    {:type :invoke, :f :txn, :value [[:append 2 4]]} 
    {:type :invoke, :f :txn, :value [[:append 0 1]]} 
    {:type :invoke, :f :txn, :value [[:r 1 nil]]} 
    {:type :invoke, :f :txn, :value [[:append 2 5]]} 
    {:type :invoke, :f :txn, :value [[:append 2 6] [:r 0 nil]]}

**Comment by happy-v587** [comment]
## 关于单元测试的说明

---

代码目录

<img width="399" alt="image" src="https://github.com/user-attachments/assets/5ca030a5-0489-42de-b18e-cf3bdd915919">

主要测试了：
- util（对parse_slots函数测试）
- file（对原生read、write、flush、get、sync的包裹）
- log（用来维护 raft log的实现）
- serialization（对 redis command 序列化与反序列化的测试）

<img width="453" alt="image" src="https://github.com/user-attachments/assets/96f836f4-0807-4d75-b3ea-4181c7a513b5">

**Comment by happy-v587** [comment]
# 一些bug处理

https://github.com/arana-db/kiwi/issues/25

**Comment by marsevilspirit** [comment]
当前时间的redis unstable无法运行redisraft, 目前使用7.2.5版本的redis可以运行

**Comment by marsevilspirit** [comment]
2025年1月18日同步进度视频：
https://pomf2.lain.la/f/nrft9b8x.MP4
